### PR TITLE
Implementation of issue #55: Implement type annotations

### PR DESCRIPTION
**Pull request by pgrandjean**

### CODE REVIEW

**Comment by joroKr21** [comment]
Thanks a lot and sorry for the delay - I will try to review this soon, but I didn't forget about it

**Comment by pgrandjean** [comment]
🙏

**Comment by pgrandjean** [comment]
rebased

File: modules/deriving/src/main/scala/shapeless3/deriving/annotation.scala [hunk 1]
```diff
@@ -124,6 +125,175 @@ object Annotations {
...
+
+/**
+ * Provides all type annotations for the fields or constructors of case class-like or sum type `T`.
+ *
+ * If type `T` is case class-like, this type class inspects its fields and provides their type annotations. If
+ * type `T` is a sum type, its constructor types are looked for type annotations as well.
+ *
+ * Type `Out` is an HList having the same number of elements as `T` (number of fields of `T` if `T` is case
+ * class-like, or number of constructors of `T` if it is a sum type). It is made of `HNil` (no annotations for corresponding
+ * field or constructor) or `HLists` (list of annotations for corresponding field or constructor).
```
Comment (joroKr21) [inline modules/deriving/src/main/scala/shapeless3/deriving/annotation.scala:248]:
```suggestion
 * Type `Out` is a tuple having the same number of elements as `T` (number of fields of `T` if `T` is a product type,
 * or number of constructors of `T` if it is a sum type). It is made of tuples
 * containing all annotations for the corresponding field or constructor (`EmptyTuple` if none).
```

File: modules/deriving/src/main/scala/shapeless3/deriving/annotation.scala [hunk 2]
```diff
@@ -124,6 +125,175 @@ object Annotations {
...
+ *
+ * Method `apply` provides an HList of type `Out` made of `HNil` (corresponding field or constructor not annotated)
+ * or `HList` (corresponding field or constructor has annotations).
```
Comment (joroKr21) [inline modules/deriving/src/main/scala/shapeless3/deriving/annotation.scala:251]:
```suggestion
 * Method `apply` provides a tuple of type `Out` made of tuples
 * containing all annotations for the corresponding field or constructor (`EmptyTuple` if none).
```

File: modules/deriving/src/main/scala/shapeless3/deriving/annotation.scala [hunk 3]
```diff
@@ -124,6 +125,175 @@ object Annotations {
...
+ *
+ * Note that type annotations must be case class-like for this type class to take them into account.
```
Comment (joroKr21) [inline modules/deriving/src/main/scala/shapeless3/deriving/annotation.scala:253]:
```suggestion
 * Note that type annotations must not be abstract for this type class to take them into account.
```

File: modules/deriving/src/main/scala/shapeless3/deriving/annotation.scala [hunk 4]
```diff
@@ -124,6 +125,175 @@ object Annotations {
...
+ *   case class First(s: String)
+ *   case class Second(i: Int)
+ *
+ *   case class CC(i: Int, @First("a") @Second(0) s: String)
+ *
+ *   val ccFirsts = AllAnnotations[CC]
+ *
+ *   // ccFirsts.Out is  ((), (First, Second))
+ *   // ccFirsts.apply() is
+ *   //   ((), (First("a"), Second(0)))
```
Comment (joroKr21) [inline modules/deriving/src/main/scala/shapeless3/deriving/annotation.scala:209]:
```suggestion
 *   // ccFirsts.Out is  (EmptyTuple, (First, Second))
 *   // ccFirsts.apply() is
 *   //   (EmptyTuple, (First("a"), Second(0)))
```

File: modules/deriving/src/main/scala/shapeless3/deriving/annotation.scala [hunk 5]
```diff
@@ -124,6 +125,175 @@ object Annotations {
...
+ *   case class First(s: String)
+ *   case class Second(i: Int)
+ *
+ *   case class CC(i: Int, s: String @First("a") @Second(0))
+ *
+ *   val ccFirsts = AllTypeAnnotations[CC]
+ *
+ *   // ccFirsts.Out is  HNil :: (First :: Second :: HNil) :: HNil
+ *   // ccFirsts.apply() is
+ *   //   HNil :: (First("a") :: Second(0) :: HNil) :: HNil
```
Comment (joroKr21) [inline modules/deriving/src/main/scala/shapeless3/deriving/annotation.scala:266]:
```suggestion
 *   // ccFirsts.Out is  (EmptyTuple, (First, Second)
 *   // ccFirsts.apply() is
 *   //   (EmptyTuple, (First("a"), Second(0))
```

File: modules/deriving/src/main/scala/shapeless3/deriving/annotation.scala [hunk 6]
```diff
@@ -145,45 +315,117 @@ object AnnotationMacros {
     }
   }
 
-  def mkAnnotations[A: Type, T: Type](using q: Quotes): Expr[Annotations[A, T]] = {
-    import quotes.reflect._
+  def mkAnnotations[A: Type, T: Type](using Quotes): Expr[Annotations[A, T]] = mkAnnotationsImpl[A, T, Annotations](ofExprVariableAnnotations[A, T](_))
```
Comment (joroKr21) [inline modules/deriving/src/main/scala/shapeless3/deriving/annotation.scala:318]:
```suggestion
  def mkAnnotations[A: Type, T: Type](using Quotes): Expr[Annotations[A, T]] = mkAnnotationsImpl[A, T, Annotations](ofExprVariableAnnotations)
```

File: modules/deriving/src/main/scala/shapeless3/deriving/annotation.scala [hunk 7]
```diff
@@ -145,45 +315,117 @@ object AnnotationMacros {
...
+
+  def mkTypeAnnotations[A: Type, T: Type](using Quotes): Expr[TypeAnnotations[A, T]] = mkAnnotationsImpl[A, T, TypeAnnotations](ofExprTypeAnnotations[A, T](_))
```
Comment (joroKr21) [inline modules/deriving/src/main/scala/shapeless3/deriving/annotation.scala:320]:
```suggestion
  def mkTypeAnnotations[A: Type, T: Type](using Quotes): Expr[TypeAnnotations[A, T]] = mkAnnotationsImpl[A, T, TypeAnnotations](ofExprTypeAnnotations)
```

File: modules/deriving/src/main/scala/shapeless3/deriving/annotation.scala [hunk 8]
```diff
@@ -145,45 +315,117 @@ object AnnotationMacros {
...
+
+  private def mkAnnotationsImpl[A: Type, T: Type, AS[A, T]: Type](mk: Seq[Expr[Any]] => Expr[AS[A, T]])(using q: Quotes): Expr[AS[A, T]] =
+    import q.reflect._
```
Comment (joroKr21) [inline modules/deriving/src/main/scala/shapeless3/deriving/annotation.scala:323]: We shouldn't use `Quotes` directly

```suggestion
  private def mkAnnotationsImpl[A: Type, T: Type, AS[A, T]: Type](mk: Seq[Expr[Any]] => Expr[AS[A, T]])(using Quotes): Expr[AS[A, T]] =
    import quotes.reflect._
```

File: modules/deriving/src/main/scala/shapeless3/deriving/annotation.scala [hunk 9]
```diff
@@ -145,45 +315,117 @@ object AnnotationMacros {
...
+
+    val tpe = TypeRepr.of[AS[A, T]] <:< TypeRepr.of[TypeAnnotations[A, T]]
+    // println(s"tpe = ${tpe}")
```
Comment (joroKr21) [inline modules/deriving/src/main/scala/shapeless3/deriving/annotation.scala:326]: Stray println

File: modules/deriving/src/main/scala/shapeless3/deriving/annotation.scala [hunk 10]
```diff
@@ -145,45 +315,117 @@ object AnnotationMacros {
...
-      def mkAnnotations(annotTrees: Seq[Expr[Any]]): Expr[Annotations[A, T]] =
-        Expr.ofTupleFromSeq(annotTrees) match {
-          case '{ $t: tup } => '{ Annotations.mkAnnotations[A, T, tup & Tuple]($t) }
-        }
-
-      def findAnnotation[A: Type](annoteeSym: Symbol): Expr[Option[A]] =
-        // TODO try to use `getAnnotation` for performance
-        annoteeSym.annotations.find(_.tpe <:< TypeRepr.of[A]) match {
+      val annotations = extractAnnotations[T](tpe)
+      // println(s"extractAnnotations = \n\t${annotations.mkString("\n\t")}")
```
Comment (joroKr21) [inline modules/deriving/src/main/scala/shapeless3/deriving/annotation.scala:334]: Please remove

File: modules/deriving/src/main/scala/shapeless3/deriving/annotation.scala [hunk 11]
```diff
@@ -145,45 +315,117 @@ object AnnotationMacros {
...
+  
+  def mkAllAnnotations[T: Type](using Quotes): Expr[AllAnnotations[T]] = mkAllAnnotationsImpl[T, AllAnnotations](ofExprAllVariableAnnotations)
+
+  def mkAllTypeAnnotations[T: Type](using Quotes): Expr[AllTypeAnnotations[T]] = mkAllAnnotationsImpl[T, AllTypeAnnotations](ofExprAllTypeAnnotations)
+
+  private def mkAllAnnotationsImpl[T: Type, AS[T]: Type](mk: Seq[Expr[Any]] => Expr[AS[T]])(using q: Quotes): Expr[AS[T]] =
+    import q.reflect._
+
+    val tpe = TypeRepr.of[AS[T]] <:< TypeRepr.of[AllTypeAnnotations[T]]
+    // println(s"tpe = ${tpe}")
```
Comment (joroKr21) [inline modules/deriving/src/main/scala/shapeless3/deriving/annotation.scala:353]: Please remove

File: modules/deriving/src/main/scala/shapeless3/deriving/annotation.scala [hunk 12]
```diff
@@ -145,45 +315,117 @@ object AnnotationMacros {
...
+
+    val annotations = extractAnnotations[T](tpe)
+    // println(s"annotations = \n\t${annotations.mkString("\n\t")}")
```
Comment (joroKr21) [inline modules/deriving/src/main/scala/shapeless3/deriving/annotation.scala:356]: Please remove

File: modules/deriving/src/main/scala/shapeless3/deriving/annotation.scala [hunk 13]
```diff
@@ -145,45 +315,117 @@ object AnnotationMacros {
...
+    mk(exprs)
+  
+  def extractAnnotations[T: Type](tpe: Boolean)(using q: Quotes): Seq[List[q.reflect.Term]] =
+    import q.reflect._
+
+    val r = new ReflectionUtils(q)
+    import r._
+
+    def getAnnotations(tree: Tree, acc: List[Term] = Nil, depth: Int = 0): List[Term] =
+      // println(s"${depth}: ${tree.show(using Printer.TreeStructure)}")
```
Comment (joroKr21) [inline modules/deriving/src/main/scala/shapeless3/deriving/annotation.scala:370]: Remove please

File: modules/deriving/src/main/scala/shapeless3/deriving/annotation.scala [hunk 14]
```diff
@@ -145,45 +315,117 @@ object AnnotationMacros {
...
+
+    mk(exprs)
+  
+  def extractAnnotations[T: Type](tpe: Boolean)(using q: Quotes): Seq[List[q.reflect.Term]] =
+    import q.reflect._
+
+    val r = new ReflectionUtils(q)
+    import r._
+
+    def getAnnotations(tree: Tree, acc: List[Term] = Nil, depth: Int = 0): List[Term] =
```
Comment (joroKr21) [inline modules/deriving/src/main/scala/shapeless3/deriving/annotation.scala:369]: Can this be `@tailrec`?

File: modules/deriving/src/main/scala/shapeless3/deriving/annotation.scala [hunk 15]
```diff
@@ -145,45 +315,117 @@ object AnnotationMacros {
...
+          case annotated: Annotated => getAnnotations(annotated.arg, annotated.annotation :: acc, depth + 1)
+          case _ => acc
+        }
+      } else {
+        tree.symbol.annotations.reverse
+      }
+      
+    @tailrec
+    def getAnnotationsFromType(typeRepr: TypeRepr, acc: List[Term] = Nil, depth: Int = 0): List[Term] =
+      // println(s"${depth}: typeRepr = ${typeRepr}")
```
Comment (joroKr21) [inline modules/deriving/src/main/scala/shapeless3/deriving/annotation.scala:386]: Remove please

File: modules/deriving/src/main/scala/shapeless3/deriving/annotation.scala [hunk 16]
```diff
@@ -145,45 +315,117 @@ object AnnotationMacros {
...
+        case typeRef: TypeRef if typeRef.typeSymbol.isAliasType => getAnnotationsFromType(typeRef.translucentSuperType, acc, depth + 1)
+        case _ => acc
       }
+
+    val annoteeTpe = TypeRepr.of[T]
+    annoteeTpe.classSymbol match {
+      case Some(annoteeCls) if annoteeCls.flags.is(Flags.Case) =>
+        val valueParams = annoteeCls.primaryConstructor
+          .paramSymss
+          .find(_.headOption.fold(false)( _.isTerm))
```
Comment (joroKr21) [inline modules/deriving/src/main/scala/shapeless3/deriving/annotation.scala:398]:
```suggestion
          .find(_.exists(_.isTerm))
```

File: modules/deriving/src/main/scala/shapeless3/deriving/annotation.scala [hunk 17]
```diff
@@ -145,45 +315,117 @@ object AnnotationMacros {
...
+          .getOrElse(Nil)
+        valueParams.map { vparam => getAnnotations(vparam.tree) }
+      case Some(annoteeCls) =>
+        Mirror(annoteeTpe) match {
+          case Some(rm) =>
+            rm.MirroredElemTypes.map { child => getAnnotations(child.typeSymbol.tree) }
+          case None =>
+            report.throwError(s"No Annotations for sum type ${annoteeTpe.show} with no Mirror")
```
Comment (joroKr21) [inline modules/deriving/src/main/scala/shapeless3/deriving/annotation.scala:406]:
```suggestion
            report.errorAndAbort(s"No Annotations for type ${annoteeTpe.show} without Mirror")
```

File: modules/deriving/src/main/scala/shapeless3/deriving/annotation.scala [hunk 18]
```diff
@@ -145,45 +315,117 @@ object AnnotationMacros {
...
+        }
+      case None =>
+        report.throwError(s"No Annotations for non-class ${annoteeTpe.show}")
```
Comment (joroKr21) [inline modules/deriving/src/main/scala/shapeless3/deriving/annotation.scala:409]:
```suggestion
        report.errorAndAbort(s"No Annotations for non-class ${annoteeTpe.show}")
```

File: modules/deriving/src/main/scala/shapeless3/deriving/annotation.scala [hunk 19]
```diff
@@ -145,45 +315,117 @@ object AnnotationMacros {
...
+    }
+
+  def ofExprVariableAnnotations[A: Type, T: Type](annotTrees: Seq[Expr[Any]])(using q: Quotes): Expr[Annotations[A, T]] =
```
Comment (joroKr21) [inline modules/deriving/src/main/scala/shapeless3/deriving/annotation.scala:412]:
```suggestion
  def ofExprVariableAnnotations[A: Type, T: Type](annotTrees: Seq[Expr[Any]])(using Quotes): Expr[Annotations[A, T]] =
```

File: modules/deriving/src/main/scala/shapeless3/deriving/annotation.scala [hunk 20]
```diff
@@ -145,45 +315,117 @@ object AnnotationMacros {
...
+    Expr.ofTupleFromSeq(annotTrees) match {
+      case '{ $t: tup } => '{ Annotations.mkAnnotations[A, T, tup & Tuple]($t) }
+    }
+
+  def ofExprTypeAnnotations[A: Type, T: Type](annotTrees: Seq[Expr[Any]])(using q: Quotes): Expr[TypeAnnotations[A, T]] =
```
Comment (joroKr21) [inline modules/deriving/src/main/scala/shapeless3/deriving/annotation.scala:417]:
```suggestion
  def ofExprTypeAnnotations[A: Type, T: Type](annotTrees: Seq[Expr[Any]])(using Quotes): Expr[TypeAnnotations[A, T]] =
```

File: modules/deriving/src/main/scala/shapeless3/deriving/annotation.scala [hunk 21]
```diff
@@ -145,45 +315,117 @@ object AnnotationMacros {
...
+    Expr.ofTupleFromSeq(annotTrees) match {
+      case '{ $t: tup } => '{ TypeAnnotations.mkAnnotations[A, T, tup & Tuple]($t) }
+    }
+
+  def ofExprAllVariableAnnotations[T: Type](annotTrees: Seq[Expr[Any]])(using q: Quotes): Expr[AllAnnotations[T]] =
```
Comment (joroKr21) [inline modules/deriving/src/main/scala/shapeless3/deriving/annotation.scala:422]:
```suggestion
  def ofExprAllVariableAnnotations[T: Type](annotTrees: Seq[Expr[Any]])(using Quotes): Expr[AllAnnotations[T]] =
```

File: modules/deriving/src/main/scala/shapeless3/deriving/annotation.scala [hunk 22]
```diff
@@ -145,45 +315,117 @@ object AnnotationMacros {
...
+    Expr.ofTupleFromSeq(annotTrees) match {
+      case '{ $t: tup } => '{ AllAnnotations.mkAnnotations[T, tup & Tuple]($t) }
+    }
+
+  def ofExprAllTypeAnnotations[T: Type](annotTrees: Seq[Expr[Any]])(using q: Quotes): Expr[AllTypeAnnotations[T]] =
```
Comment (joroKr21) [inline modules/deriving/src/main/scala/shapeless3/deriving/annotation.scala:427]:
```suggestion
  def ofExprAllTypeAnnotations[T: Type](annotTrees: Seq[Expr[Any]])(using Quotes): Expr[AllTypeAnnotations[T]] =
```

File: modules/deriving/src/test/scala/shapeless3/deriving/annotation.scala [hunk 23]
```diff
@@ -120,4 +149,81 @@ class AnnotationTests {
...
+      val secondSum = TypeAnnotations[Second, Base2].apply()
+      assert(secondSum == (None, Some(Second(3, "e"))))
+    }
+  }
+
+  @Test
+  def invalidTypeAnnotations: Unit = {
+    illTyped(" TypeAnnotations[Dummy, CC2] ", "could not find implicit value for parameter annotations: .*")
+    illTyped(" TypeAnnotations[Dummy, Base] ", "could not find implicit value for parameter annotations: .*")
+    illTyped(" TypeAnnotations[Second, Dummy] ", "could not find implicit value for parameter annotations: .*")
```
Comment (joroKr21) [inline modules/deriving/src/test/scala/shapeless3/deriving/annotation.scala:194]:
```suggestion
    illTyped("TypeAnnotations[Dummy, CC2]", "could not find implicit value for parameter annotations: .*")
    illTyped("TypeAnnotations[Dummy, Base]", "could not find implicit value for parameter annotations: .*")
    illTyped("TypeAnnotations[Second, Dummy]", "could not find implicit value for parameter annotations: .*")
```

File: modules/deriving/src/test/scala/shapeless3/deriving/annotation.scala [hunk 24]
```diff
@@ -120,4 +149,81 @@ class AnnotationTests {
...
+  }
+
+  @Test
+  def allAnnotations: Unit = {
+    type T1First = First *: EmptyTuple.type
```
Comment (joroKr21) [inline modules/deriving/src/test/scala/shapeless3/deriving/annotation.scala:199]:
```suggestion
    type T1First = First *: EmptyTuple
```

File: modules/deriving/src/test/scala/shapeless3/deriving/annotation.scala [hunk 25]
```diff
@@ -120,4 +149,81 @@ class AnnotationTests {
...
+    val first: T1First = First() *: EmptyTuple
+
+    val cc: (T1First, EmptyTuple.type, (Second, Third)) = AllAnnotations[CC3].apply()
```
Comment (joroKr21) [inline modules/deriving/src/test/scala/shapeless3/deriving/annotation.scala:202]:
```suggestion
    val cc: (T1First, EmptyTuple, (Second, Third)) = AllAnnotations[CC3].apply()
```

File: modules/deriving/src/test/scala/shapeless3/deriving/annotation.scala [hunk 26]
```diff
@@ -120,4 +149,81 @@ class AnnotationTests {
...
+    assert(cc == (first, EmptyTuple, (Second(2, "b"), Third('c'))))
+
+    type T1Second = Second *: EmptyTuple.type
```
Comment (joroKr21) [inline modules/deriving/src/test/scala/shapeless3/deriving/annotation.scala:205]:
```suggestion
    type T1Second = Second *: EmptyTuple
```

File: modules/deriving/src/test/scala/shapeless3/deriving/annotation.scala [hunk 27]
```diff
@@ -120,4 +149,81 @@ class AnnotationTests {
...
+    val second: T1Second = Second(3, "e") *: EmptyTuple
+
+    val st: (T1First, T1Second) = AllAnnotations[Base].apply()
+    assert(st == (first, second))
+  }
+
+  @Test
+  def allTypeAnnotations: Unit = {
+    type T1First = First *: EmptyTuple.type
```
Comment (joroKr21) [inline modules/deriving/src/test/scala/shapeless3/deriving/annotation.scala:214]:
```suggestion
    type T1First = First *: EmptyTuple
```

File: modules/deriving/src/test/scala/shapeless3/deriving/annotation.scala [hunk 28]
```diff
@@ -120,4 +149,81 @@ class AnnotationTests {
...
+    val first: T1First = First() *: EmptyTuple
+
+    val st: (T1First, (Second, Third)) = AllTypeAnnotations[Base2].apply() // sealed trait
+    assert(st == (first, (Second(3, "e"), Third('c'))))
+
+    val cc: (T1First, EmptyTuple.type, (Second, Third)) = AllTypeAnnotations[CC4].apply() // case class
```
Comment (joroKr21) [inline modules/deriving/src/test/scala/shapeless3/deriving/annotation.scala:220]:
```suggestion
    val cc: (T1First, EmptyTuple, (Second, Third)) = AllTypeAnnotations[CC4].apply() // case class
```

File: modules/deriving/src/test/scala/shapeless3/deriving/annotation.scala [hunk 29]
```diff
@@ -120,4 +149,81 @@ class AnnotationTests {
...
+    assert(cc == (first, EmptyTuple, (Second(2, "b"), Third('c'))))
+
+    type T1Third = Third *: EmptyTuple.type
```
Comment (joroKr21) [inline modules/deriving/src/test/scala/shapeless3/deriving/annotation.scala:223]:
```suggestion
    type T1Third = Third *: EmptyTuple
```

**Review by joroKr21 — CHANGES_REQUESTED** [review]
Sorry for taking so long to review 🙈 

Overall looks good, the documentation can be improved and some polishing to do.
We should also add some tests with enums - that's the future of ADTs.

File: modules/deriving/src/main/scala/shapeless3/deriving/annotation.scala (hunk 14 above)
Comment (joroKr21) [reply modules/deriving/src/main/scala/shapeless3/deriving/annotation.scala:369]: Looks like it can't

**Review by joroKr21 — APPROVED** [review]
I addressed my comments - we still need to add more tests but this PR is getting too big.

**Comment by pgrandjean** [comment]
@joroKr21 thank you for all the fixes and integration!

**Comment by pgrandjean** [comment]
@joroKr21 what tests need to be added?

**Comment by joroKr21** [comment]
> @joroKr21 what tests need to be added?

We were lacking tests with enums - I added one though

**Comment by pgrandjean** [comment]
I tried using type annotations on enums, but the following syntax is invalid:
```scala
enum Control2:
  case Automatic @First
  case Manual(age: PosInt, email: Email) @Third('!')
```
according to Scala's [EBNF](https://dotty.epfl.ch/docs/internals/syntax.html), the best I can do is:
```scala
enum Control2:
  case Automatic extends Control2 @First
  case Manual(age: PosInt, email: Email) extends Control2 @Third('!')
```
which kills the purpose of the new enum syntax by repeating `extends Control2`. If you see an interest in supporting this syntax, I can add a test.

**Comment by joroKr21** [comment]
@pgrandjean this is the test I added:
```scala
@Other @Last(false) enum Control:
    @First @Third('!') case Automatic
    @Second(100, "@") case Manual(age: PosInt, email: Email)
```

Not sure if that corresponds to your expectations. I think you know better than me how this functionality is supposed to be used. More tests can't hurt, but at least we already have one test for sanity check, so if you want to add more it's up to you.

Btw this is not released yet. I got a bit stuck on #36 - wanted to complete it for the next release but it looks still half-baked. I'm thinking of a more useful design so I will leave if for another milestone.

**Pull request merged**